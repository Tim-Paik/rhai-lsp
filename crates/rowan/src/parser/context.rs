//! The parser context is a separate module to limit
//! the API surface for the parser functions.
#![allow(dead_code)]
use rowan::{GreenNodeBuilder, TextRange, TextSize};

use crate::syntax::{
    Lexer,
    SyntaxKind::{self, *},
};

use super::{Parse, ParseError, ParseErrorKind};

pub(crate) struct Context<'src> {
    lexer: Lexer<'src>,
    current_token: Option<SyntaxKind>,
    last_token: Option<SyntaxKind>,
    green: GreenNodeBuilder<'static>,
    errors: Vec<ParseError>,

    /// Tracks statements being separated by ";".
    statement_closed: bool,
}

impl<'src> Context<'src> {
    pub(crate) fn new(source: &'src str) -> Self {
        Self {
            lexer: Lexer::new(source),
            current_token: None,
            last_token: None,
            green: GreenNodeBuilder::new(),
            errors: Vec::new(),

            statement_closed: true,
        }
    }

    pub(crate) fn token(&mut self) -> Option<SyntaxKind> {
        // Eat insignificant tokens
        loop {
            if self.current_token.is_none() {
                self.current_token = self.lexer.next();
            }

            match self.current_token {
                Some(COMMENT_BLOCK) | Some(COMMENT_LINE) | Some(WHITESPACE) => {
                    self.eat();
                    self.last_token = None // We aren't interested in these.
                }
                Some(ERROR) => {
                    self.eat_error(ParseErrorKind::InvalidInput);
                    self.last_token = None // We aren't interested in these.
                }
                _ => break,
            }
        }

        self.current_token
    }

    pub(crate) fn previous_token(&self) -> Option<SyntaxKind> {
        self.last_token
    }

    pub(crate) fn eat(&mut self) {
        if let Some(t) = self.current_token.take() {
            self.green.token(t.into(), self.lexer.slice());
            self.last_token = Some(t);
        }
        self.current_token = None;
    }

    pub(crate) fn eat_as(&mut self, kind: SyntaxKind) {
        if self.current_token.is_some() {
            self.green.token(kind.into(), self.lexer.slice());
            self.last_token = Some(kind);
        }
        self.current_token = None;
    }

    pub(crate) fn eat_error(&mut self, error: ParseErrorKind) {
        self.add_error(error);
        self.eat();
    }

    pub(crate) fn eat_error_as(&mut self, kind: SyntaxKind, error: ParseErrorKind) {
        self.add_error(error);
        self.eat_as(kind);
    }

    pub(crate) fn add_error(&mut self, error: ParseErrorKind) {
        let span = self.lexer.span();
        self.errors.push(ParseError::new(
            TextRange::new(
                TextSize::from(span.start as u32),
                TextSize::from(span.end as u32),
            ),
            error,
        ));
    }

    pub(crate) fn start_node(&mut self, kind: SyntaxKind) {
        self.green.start_node(kind.into());
    }

    pub(crate) fn finish_node(&mut self) {
        self.green.finish_node();
    }

    pub(crate) fn finish(self) -> Parse {
        Parse {
            errors: self.errors,
            green: self.green.finish(),
        }
    }

    pub(crate) fn statement_closed(&self) -> bool {
        self.statement_closed
    }

    pub(crate) fn set_statement_closed(&mut self, v: bool) {
        self.statement_closed = v;
    }
}
