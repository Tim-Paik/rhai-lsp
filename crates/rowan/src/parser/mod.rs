//! This module contains all parsing-related tools including a
//! recursive-descent parser.

use crate::syntax::{SyntaxKind, SyntaxNode};
use rowan::GreenNode;
use thiserror::Error;

use self::context::Context;

mod context;
pub mod parsers;

/// A flexible parser.
/// 
/// Parsing happens via given parser functions (also found in [`parsers`]).
/// 
/// # Example Uses
/// 
/// ```
/// use rhai_rowan::parser::Parser;
/// // Parse some Rhai code (file).
/// let parser = Parser::new(r#"print("hello rhai!")"#);
/// let parse = parser.parse();
/// 
/// assert!(parse.errors.is_empty());
/// ```
/// 
/// ```
/// use rhai_rowan::parser::Parser;
/// use rhai_rowan::parser::parsers::parse_expr;
///
/// // Parse an expression explicitly.
/// let mut parser = Parser::new(r#"print("hello rhai!")"#);
/// parser.execute(parse_expr);
/// let parse = parser.finish();
/// 
/// assert!(parse.errors.is_empty());
/// ```
pub struct Parser<'src> {
    context: Context<'src>,
}

impl<'src> Parser<'src> {
    /// Create a new parser with the given source.
    #[must_use]
    pub fn new(source: &'src str) -> Self {
        Self {
            context: Context::new(source),
        }
    }

    /// Finish parsing.
    /// 
    /// # Panics
    /// 
    /// If no parser function was called,
    /// or a parser was left in an invalid state (bug).
    #[must_use]
    pub fn finish(self) -> Parse {
        self.context.finish()
    }
}

impl<'src> Parser<'src> {
    /// Execute a parser function for the given parser.
    pub fn execute<F: FnOnce(&mut Context)>(&mut self, f: F) {
        f(&mut self.context);
    }
}

/// The result of parsing, containing all errors
/// and the final green tree.
#[derive(Debug, Clone)]
pub struct Parse {
    /// Parse errors.
    pub errors: Vec<ParseError>,
    /// Parsed green tree.
    pub green: GreenNode,
}

impl Parse {
    /// Turn the result green tree into a CST.
    /// *This ignores errors*, the resulting tree
    /// can be potentially syntactically invalid.
    #[must_use]
    pub fn clone_syntax(&self) -> SyntaxNode {
        SyntaxNode::new_root(self.green.clone())
    }

    /// Turn the result green tree into a CST.
    /// *This ignores errors*, the resulting tree
    /// can be potentially syntactically invalid.
    #[must_use]
    pub fn into_syntax(self) -> SyntaxNode {
        SyntaxNode::new_root(self.green)
    }
}

/// A parse (syntax) error.
#[derive(Debug, Clone, PartialEq, Eq, Error)]
#[error("parse error at {range:?}: {kind}")]
pub struct ParseError {
    /// The span of the error in the parsed source.
    pub range: rowan::TextRange,
    /// Error kind.
    pub kind: ParseErrorKind,
}

impl ParseError {
    fn new(range: rowan::TextRange, kind: ParseErrorKind) -> Self {
        Self { range, kind }
    }
}

/// All the non-fatal parsing errors that can occur.
#[derive(Debug, PartialEq, Eq, Clone, Error)]
pub enum ParseErrorKind {
    #[error("unexpected EOF")]
    UnexpectedEof,

    #[error(r#"invalid input"#)]
    InvalidInput,

    #[error(r#"unexpected token"#)]
    UnexpectedToken,

    #[error(r#"expected token "{0:?}""#)]
    ExpectedToken(SyntaxKind),

    #[error(
        r#"expected one of "{}""#,
        .0
            .iter()
            .map(|s|format!("{:?}", s))
            .collect::<Vec<String>>().join(",")
    )]
    ExpectedOneOfTokens(Vec<SyntaxKind>),
}
