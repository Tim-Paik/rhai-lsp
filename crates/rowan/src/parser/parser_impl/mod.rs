#![deny(unreachable_patterns)]

use super::context::Context;
use crate::parser::ParseErrorKind;
use crate::syntax::SyntaxKind::*;
use crate::T;

macro_rules! require_token {
    ($ctx:ident) => {
        match $ctx.token() {
            Some(t) => t,
            None => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
        }
    };

    ($ctx:ident, finish_node) => {
        match $ctx.token() {
            Some(t) => t,
            None => {
                $ctx.finish_node();
                $ctx.eat_error(ParseErrorKind::UnexpectedEof);
                return;
            }
        }
    };
}

macro_rules! expect_token {
    ($ctx:ident, finish_node, $($token:tt)*) => {
        match $ctx.token() {
            Some(t) => match t {
                $($token)* => {
                    $ctx.eat();
                },
                _ => {
                    $ctx.finish_node();
                    $ctx.eat_error(ParseErrorKind::ExpectedToken($($token)*));
                    return;
                }
            }
            None => {
                $ctx.finish_node();
                $ctx.eat_error(ParseErrorKind::ExpectedToken($($token)*));
                return;
            }
        }
    };
    ($ctx:ident, $($token:tt)+) => {
        match $ctx.token() {
            Some(t) => match t {
                $($token)* => {
                    $ctx.eat();
                },
                _ => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
            }
            None => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
        }
    };

}

impl<'src> super::Parser<'src> {
    pub fn parse_file(&mut self) {
        parse_file(&mut self.context);
    }
}

fn parse_file(ctx: &mut Context) {
    ctx.start_node(FILE);
    if let Some(SHEBANG) = ctx.token() {
        parse_shebang(ctx);
    }

    ctx.set_statement_closed(true);
    while ctx.token().is_some() {
        if !ctx.statement_closed() {
            ctx.add_error(ParseErrorKind::ExpectedToken(T![";"]));
        }

        parse_stmt(ctx);
    }
    ctx.set_statement_closed(true);

    ctx.finish_node()
}

fn parse_shebang(ctx: &mut Context) {
    let token = require_token!(ctx);

    if !matches!(token, SHEBANG) {
        return ctx.eat_error(ParseErrorKind::ExpectedToken(SHEBANG));
    }

    ctx.eat()
}

fn parse_stmt(ctx: &mut Context) {
    let token = require_token!(ctx);
    ctx.start_node(STMT);
    ctx.set_statement_closed(false);

    if token == T![";"] {
        ctx.finish_node();
        ctx.set_statement_closed(true);
        return ctx.eat();
    }

    parse_item(ctx);

    if let Some(token) = ctx.token() {
        if token == T![";"] {
            ctx.eat();
            ctx.set_statement_closed(true);
        }
    }

    ctx.finish_node();
}

fn parse_item(ctx: &mut Context) {
    ctx.start_node(ITEM);

    // Parse doc comments if any.
    while matches!(
        require_token!(ctx, finish_node),
        COMMENT_BLOCK_DOC | COMMENT_LINE_DOC
    ) {
        ctx.eat();
    }

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr(ctx: &mut Context) {
    ctx.start_node(EXPR);
    let token = require_token!(ctx, finish_node);

    match token {
        T!["let"] => parse_expr_let(ctx),
        T!["const"] => parse_expr_const(ctx),
        T!["{"] => parse_expr_block(ctx),
        T!["fn"] => parse_expr_fn(ctx),
        LIT_INT | LIT_FLOAT | LIT_BOOL | LIT_STR | LIT_CHAR => parse_expr_lit(ctx),
        IDENT => parse_expr_ident(ctx),
        _ => ctx.eat_error(ParseErrorKind::UnexpectedToken),
    }

    ctx.finish_node();
}

fn parse_expr_ident(ctx: &mut Context) {
    ctx.start_node(EXPR_IDENT);
    parse_ident(ctx);
    ctx.finish_node();
}

fn parse_expr_lit(ctx: &mut Context) {
    ctx.start_node(EXPR_LIT);
    parse_lit(ctx);
    ctx.finish_node();
}

fn parse_expr_let(ctx: &mut Context) {
    ctx.start_node(EXPR_LET);

    expect_token!(ctx, finish_node, T!["let"]);
    expect_token!(ctx, finish_node, T!["ident"]);
    expect_token!(ctx, finish_node, T!["="]);

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr_const(ctx: &mut Context) {
    ctx.start_node(EXPR_CONST);

    expect_token!(ctx, finish_node, T!["const"]);
    expect_token!(ctx, finish_node, T!["ident"]);
    expect_token!(ctx, finish_node, T!["="]);

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr_block(ctx: &mut Context) {
    ctx.start_node(EXPR_BLOCK);

    expect_token!(ctx, finish_node, T!["{"]);

    ctx.set_statement_closed(true);
    loop {
        let token = require_token!(ctx, finish_node);

        if token == T!["}"] {
            break;
        }

        if !ctx.statement_closed() {
            ctx.add_error(ParseErrorKind::ExpectedToken(T![";"]));
        }

        parse_stmt(ctx);
    }

    expect_token!(ctx, finish_node, T!["}"]);

    // Blocks also don't require ";" when used as statements.
    ctx.set_statement_closed(true);

    ctx.finish_node();
}

fn parse_expr_fn(ctx: &mut Context) {
    ctx.start_node(EXPR_FN);

    expect_token!(ctx, finish_node, T!["fn"]);
    expect_token!(ctx, finish_node, T!["ident"]);

    parse_param_list(ctx);
    parse_expr_block(ctx);

    ctx.finish_node();
}

fn parse_param_list(ctx: &mut Context) {
    ctx.start_node(PARAM_LIST);

    let start_token = require_token!(ctx, finish_node);

    match start_token {
        T!["("] | T!["|"] => {
            ctx.eat();
        }
        _ => {
            ctx.eat_error(ParseErrorKind::ExpectedOneOfTokens(vec![T!["("], T!["|"]]));
            return;
        }
    }

    loop {
        let token = require_token!(ctx, finish_node);
        if matches!(
            (start_token, token),
            (T!["("], T![")"]) | (T!["|"], T!["|"])
        ) {
            ctx.eat();
            break;
        }

        parse_param(ctx);

        let end_token = require_token!(ctx, finish_node);

        match (start_token, end_token) {
            (T!["("], T![")"]) | (T!["|"], T!["|"]) => {
                ctx.eat();
                break;
            }
            (_, T![","]) => {
                ctx.eat();
            }
            _ => {
                ctx.eat_error(ParseErrorKind::ExpectedToken(T![","]));
                break;
            }
        }
    }

    ctx.finish_node();
}

fn parse_param(ctx: &mut Context) {
    ctx.start_node(PARAM);

    expect_token!(ctx, finish_node, T!["ident"]);

    ctx.finish_node();
}

fn parse_lit(ctx: &mut Context) {
    ctx.start_node(LIT);
    let token = require_token!(ctx, finish_node);

    match token {
        LIT_INT | LIT_FLOAT | LIT_BOOL | LIT_STR | LIT_CHAR => {
            ctx.eat();
        }
        _ => {
            ctx.eat_error(ParseErrorKind::ExpectedOneOfTokens(vec![
                LIT_INT, LIT_FLOAT, LIT_BOOL, LIT_STR, LIT_CHAR,
            ]));
        }
    }

    ctx.finish_node();
}

fn parse_ident(ctx: &mut Context) {
    let token = require_token!(ctx);

    match token {
        IDENT => ctx.eat(),
        _ => ctx.eat_error(ParseErrorKind::ExpectedToken(IDENT)),
    }
}

// Expr =
// | ExprUni
// | ExprBin
// | ExprParen
// | ExprArray
// | ExprIndex
// | ExprObject
// | ExprCall
// | ExprMethodCall
// | ExprAccess
// | ExprClosure
// | ExprIf
// | ExprLoop
// | ExprFor
// | ExprWhile
// | ExprBreak
// | ExprContinue
// | ExprSwitch
// | ExprReturn
// | ExprPath
// | ExprImport
