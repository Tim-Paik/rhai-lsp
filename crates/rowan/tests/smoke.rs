use rhai_rowan::parser::Parser;
use test_case::test_case;

#[test_case(include_str!("../../../testdata/valid/simple.rhai"))]
#[test_case(include_str!("../../../testdata/valid/oop.rhai"))]
#[test_case(include_str!("../../../testdata/valid/module-example.rhai"))]
fn parse_valid(src: &str) {
    let parse = Parser::new(src).parse();
    assert!(parse.errors.is_empty());
    insta::assert_snapshot!(format!("{:#?}", parse.into_syntax()));
}

// #[test_case(include_str!("../../../testdata/valid/module-example.d.rhai"))]
// #[test_case(include_str!("../../../testdata/valid/module-global.d.rhai"))]
// #[test_case(include_str!("../../../testdata/valid/module-named.d.rhai"))]
// #[test_case(include_str!("../../../testdata/valid/module-static.d.rhai"))]
#[test]
fn parse_valid_definitions() {
    let parse = Parser::new(include_str!("../../../testdata/valid/module-static.d.rhai")).parse_def();
    assert!(parse.errors.is_empty());
    insta::assert_snapshot!(format!("{:#?}", parse.into_syntax()));
}
