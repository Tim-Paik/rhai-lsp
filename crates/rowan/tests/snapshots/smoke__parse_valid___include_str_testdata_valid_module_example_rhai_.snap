---
source: crates/rowan/tests/smoke.rs
expression: "format!(\"{:#?}\", parse.into_syntax())"

---
RHAI@0..86
  STMT@0..34
    ITEM@0..32
      EXPR@0..32
        EXPR_FN@0..32
          KW_FN@0..2 "fn"
          WHITESPACE@2..3 " "
          IDENT@3..8 "hello"
          PARAM_LIST@8..14
            PUNCT_PAREN_START@8..9 "("
            PARAM@9..13
              IDENT@9..13 "name"
            PUNCT_PAREN_END@13..14 ")"
          EXPR_BLOCK@14..32
            WHITESPACE@14..15 " "
            PUNCT_BRACE_START@15..16 "{"
            WHITESPACE@16..21 "\n    "
            STMT@21..30
              ITEM@21..29
                EXPR@21..29
                  EXPR_CALL@21..29
                    EXPR@21..23
                      EXPR_IDENT@21..23
                        IDENT@21..23 "hi"
                    ARG_LIST@23..29
                      PUNCT_PAREN_START@23..24 "("
                      EXPR@24..28
                        EXPR_IDENT@24..28
                          IDENT@24..28 "name"
                      PUNCT_PAREN_END@28..29 ")"
              PUNCT_SEMI@29..30 ";"
            WHITESPACE@30..31 "\n"
            PUNCT_BRACE_END@31..32 "}"
    WHITESPACE@32..34 "\n\n"
  STMT@34..86
    ITEM@34..85
      EXPR@34..85
        EXPR_FN@34..85
          KW_PRIVATE@34..41 "private"
          WHITESPACE@41..42 " "
          KW_FN@42..44 "fn"
          WHITESPACE@44..45 " "
          IDENT@45..47 "hi"
          PARAM_LIST@47..53
            PUNCT_PAREN_START@47..48 "("
            PARAM@48..52
              IDENT@48..52 "name"
            PUNCT_PAREN_END@52..53 ")"
          EXPR_BLOCK@53..85
            WHITESPACE@53..54 " "
            PUNCT_BRACE_START@54..55 "{"
            WHITESPACE@55..60 "\n    "
            STMT@60..83
              ITEM@60..82
                EXPR@60..82
                  EXPR_CALL@60..82
                    EXPR@60..65
                      EXPR_IDENT@60..65
                        IDENT@60..65 "print"
                    ARG_LIST@65..82
                      PUNCT_PAREN_START@65..66 "("
                      EXPR@66..81
                        EXPR_LIT@66..81
                          LIT@66..81
                            LIT_STR@66..81 "`hello ${name}`"
                      PUNCT_PAREN_END@81..82 ")"
              PUNCT_SEMI@82..83 ";"
            WHITESPACE@83..84 "\n"
            PUNCT_BRACE_END@84..85 "}"
    WHITESPACE@85..86 "\n"

